### PR TITLE
Shared project module clipboard

### PR DESCRIPTION
**Pull request by GantMan**
The .NET46 clipboard 👍 

Structure is similar to UWP, even some items are there as a facade, but most importantly it's geared around async read/write for actual use.

Ended up using: https://github.com/undergroundwires/AsyncWindowsClipboard

### CODE REVIEW

File: ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj [hunk 1]
```diff
@@ -52,6 +52,10 @@
     </DefineConstants>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="AsyncClipboardService, Version=1.5.0.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\AsyncClipboardService.1.5.0\lib\net452\AsyncClipboardService.dll</HintPath>
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj]: Seeing quite a few of these small dependencies.  Is it that much harder to use the WPF interfaces directly for clipboard?

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace ReactNative.Modules.Clipboard
+{
+    /// <summary>
+    /// Gets and sets information from the clipboard object.
+    /// </summary>
+    public interface IClipboardInstance
+    {
+        /// <summary>
+        /// Removes all data from the Clipboard.
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs]: Inconsistent use of Clipboard vs. clipboard

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        /// <summary>
+        /// Gets the current content that is stored in the clipboard object.
+        /// </summary>
+        /// <returns>Contains the content of the Clipboard.</returns>
+        IDataObject GetContent();
+
+        /// <summary>
+        /// Sets the current content that is stored in the clipboard object.
+        /// </summary>
+        /// <param name="package">Contains the content of the clipboard. If NULL, the clipboard is emptied.</param>
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs]: NULL -> <code>null</code>

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace ReactNative.Modules.Clipboard
+{
+    /// <summary>
+    /// A module that allows JS to get/set clipboard contents.
+    /// </summary>
+    class ClipboardModule : NativeModuleBase
+    {
+        private readonly IClipboardInstance _clipboard;
+        private WindowsClipboardService _clipboardService = new WindowsClipboardService();
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs]: Not readonly?

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
+﻿using ReactNative.Bridge;
+using System;
+using System.Threading.Tasks;
+using AsyncClipboardService.Clipboard;
+using AsyncWindowsClipboard;
+using System.Threading;
+using System.Windows.Threading;
+// using DataTransfer = Windows.ApplicationModel.DataTransfer;
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs]: delete unused namespace

File: ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj (hunk 1 above)
Comment (kevinvangelder) [inline ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj]: Gant told me earlier today that the WPF interface doesn't support async clipboard access.

Comment (rozele) [inline ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj]: Does it need to be async?

**Comment by GantMan** [comment]
### @rozele -  Quick update
* Removed the nuget dependency
* Does not use async, but clipboard doesn't have to (as you asked)
* Documented/Cleaned up code noise

File: ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj [hunk 6]
```diff
@@ -92,13 +92,17 @@
...
+    <Compile Include="Modules\Clipboard\IClipboardInstance.cs" />
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="Tracing\NullLoggingActivityBuilder.cs" />
     <Compile Include="Tracing\Tracer.cs" />
   </ItemGroup>
   <ItemGroup>
     <None Include="packages.config" />
   </ItemGroup>
+  <ItemGroup />
   <Import Project="..\ReactNative.Shared\ReactNative.Shared.projitems" Label="Shared" />
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/ReactNative.Net46.csproj]: Could delete if you want.

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+        /// <summary>
+        /// Add text to the clipboard or clear the clipboard.
+        /// </summary>
+        /// <param name="text">The text. If null clear clipboard.</param>
+        [ReactMethod]
+        public void setString(string text)
+        {
+            _clipboard.SetText(text);
+        }
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/ClipboardModule.cs]: Extra newline after closing curly should be removed.

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs (hunk 3 above)
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs]: lowercase null

---
In reply to: [82703726](https://github.com/ReactWindows/react-native-windows/pull/759#discussion_r82703726) [](ancestors = 82703726)

File: ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        void SetContent(IDataObject package);
+
+        bool ContainsText();
```
Comment (rozele) [inline ReactWindows/ReactNative.Net46/Modules/Clipboard/IClipboardInstance.cs]: Missing comment for the returns block here and below.

---
In reply to: [82703640](https://github.com/ReactWindows/react-native-windows/pull/759#discussion_r82703640) [](ancestors = 82703640)

**Comment by rozele** [comment]
This looks good, mostly nitpick comments remaining but it would be good to clean them up.  We really do need to get some kind of code analysis rules running so you can "self-help" ;-)

---
In reply to: [252960235](https://github.com/ReactWindows/react-native-windows/pull/759#issuecomment-252960235) [](ancestors = 252960235)

**Comment by GantMan** [comment]
@rozele - suggestions integrated.

**Comment by GantMan** [comment]
@rozele - rebased back

**Comment by rozele** [comment]
Needs rebase and then good to go.

**Pull request merged**